### DISCUSSION TITLE
Is it possible to disable SSR/SSG completely in Next.js?

### DISCUSSION THREAD

**Discussion by josh-bitovi** [vercel/next.js#22430] [Help]
Cross-posting [this question](https://stackoverflow.com/questions/66322077/is-it-possible-to-disable-ssr-ssg-completely-in-next-js) here for visibility. Thanks for any 👀 !

**Comment by michrome** [comment]
[`next export`](https://nextjs.org/docs/advanced-features/static-html-export) allows you to export your app to static HTML, which can be run standalone without the need of a Node.js server. You could host its output on S3.

**Reply by josh-bitovi** [reply]
Right, I'm using that, I'm just wondering if I can disable things like `getServerSideProps` _at build time_, so I can keep them in the code and use them in the SSR version of the app (but _not_ generate a SSG app-- instead generate a completely static app with `next export`).

**Reply by michrome** [reply]
That's _sort of_ how [Automatic Static Optimization](https://nextjs.org/docs/advanced-features/automatic-static-optimization) works:

* If a page contains `getServerSideProps` or `getInitialProps` then it will be Server-Side Rendered: `next build` will create a **JavaScript** output like `.next/server/pages/ssr-page.js`.
* If a page contains `getStaticProps` then it will be Statically Generated: `next build` will create an **HTML** output like `.next/server/pages/ssg-page.html`.

But I think what you're looking for is something like https://github.com/vercel/next.js/discussions/17271

**Reply by josh-bitovi** [reply]
Right, I'm looking for the static optimization-- but I'd like to choose explicitly at build time, so I can have an SSR app, and a fully static app-- effectively stripping out `getServerSideProps`/`getInitialProps` at build time, but only for the static build.

I'm not sure https://github.com/vercel/next.js/discussions/17271 is it-- I'd still like the pages... I just want them to be `/posts/[post].html` rather than that + some json files (SSG).

**Reply by michrome** [reply]
Grasping at straws … would `unstable_runtimeJS` help?

If you had this:

```js
// pages/[id].js
export const config = {
  unstable_runtimeJS: false,
};

function Page({ id }) {
  return <h1>Page {id}</h1>;
}

export async function getStaticPaths() {
  return {
    paths: [{ params: { id: "1" } }, { params: { id: "2" } }],
    fallback: false,
  };
}

export async function getStaticProps({ params }) {
  return { props: { id: params.id } };
}

export default Page;
```

Then the output of `next export` would include the following. You'd lose all your Next/React/JS goodness.

```html
<!-- out/1.html -->
<!DOCTYPE html>
<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width" />
    <meta name="next-head-count" content="2" />
    <noscript data-n-css=""></noscript>
  </head>
  <body>
    <div id="__next">
      <h1>
        Page
        <!-- -->1
      </h1>
    </div>
  </body>
</html>
```

PS - You will get a `out/1.html` from this example, even without using `unstable_runtimeJS `.

**Reply by josh-bitovi** [reply]
That's so close! Reading through that thread it seems like their goal is _no runtime JS_... however my goal is slightly different... I just want static (without prerendered paths/data) pages. For example if I build a page that uses `unstable_runtimeJS`, I get:
```
└ ● /products/[name]             1.76 kB        65.3 kB
    ├ /products/apples
    ├ /products/oranges
    └ /products/bananas
...

λ  (Server)  server-side renders at runtime (uses getInitialProps or getServerSideProps)
○  (Static)  automatically rendered as static HTML (uses no initial props) // 🥅
●  (SSG)     automatically generated as static HTML + JSON (uses getStaticProps)
   (ISR)     incremental static regeneration (uses revalidate in getStaticProps)
```
But if I _completely remove_ `getStaticProps`/`getStaticPaths`/`getServerSideProps`, I get:
```
└ ○ /products/[name]             1.76 kB        65.3 kB (different sizes though) // ✅
```

**Reply by michrome** [reply]
The difference between `○  (Static)` and `●  (SSG)` comes down to whether or not your page has any initial `props`.

If your page doesn't have any initial props, then you don't need `getStaticProps`. You also don't need `getStaticPaths` because—without any `props`—`/apples`, `/oranges` and `/bananas` would all have **identical** content. As you've seen, if you don't need these methods, you'll get one `○  (Static)` page.

If your page does have props (presumably meaning `/apples`, `/oranges` and `/bananas` do have different content), you'll get `●  (SSG)` pages. `● (SSG)` pages are static HTML:

 * You have one html file for each of your static paths (`apples.html`, `oranges.html`, `bananas.html`, etc)
 * You can host them on S3
 * You can `curl` each html file and you'll see its props already in place (unlike Client-Side rendered pages)
 * Statically generated pages are still reactive: Next.js will hydrate your application client-side to give it full interactivity—that's what the additional JSON files are providing.

**Reply by josh-bitovi** [reply]
I know, we're just trying to avoid a (possibly large) extra build step that we might retain on the SSR side of things.

Appreciate all the discussion on this. It seems like there isn't a [next.config.js](https://nextjs.org/docs/api-reference/next.config.js/introduction) setting for this... so I'll try to build them out if I can.
